pkg/controller/metrics/metricsWithDynamicLabels.go

```diff
@@ -0,0 +1,91 @@
+func (c CounterWithDynamicLabels) LogMetricWithDynamicLabels(obj metav1.Object, log log.FieldLogger, _ float64) {
+	curriedVec, err := c.CounterVec.CurryWith(c.FixedLabels)
+	// Maps cannot be sorted, however we would want to loop through them in a sorted fashion to ensure all the metric
+	// labels are added in an expected order. Get the slice of map keys first, sort them and then use it to loop over
+	// the map
+	for _, v := range GetSortedKeys(c.OptionalLabels) {
+		curriedVec, err = curriedVec.CurryWith(prometheus.Labels{v: GetLabelValue(obj, c.OptionalLabels[v])})
+	}
+	if err != nil {
+		// The possibility of hitting an error here is low, but report it and don't observe the metric in case it happens
+		log.WithError(err).Error("error observing metric")
+	} else {
+		curriedVec.WithLabelValues().Inc()
+	}
+}
+
+type HistogramWithDynamicLabels struct {
+	*prometheus.HistogramVec
+	FixedLabels    map[string]string
+	OptionalLabels map[string]string
+}
+
+// LogMetricWithDynamicLabels observes the histogram metric with val value.
+func (h HistogramWithDynamicLabels) LogMetricWithDynamicLabels(obj metav1.Object, log log.FieldLogger, val float64) {
+	curriedVec, err := h.HistogramVec.CurryWith(h.FixedLabels)
+	// Maps cannot be sorted, however we would want to loop through them in a sorted fashion to ensure all the metric
+	// labels are added in an expected order. Get the slice of map keys first, sort them and then use it to loop over
+	// the map
+	for _, v := range GetSortedKeys(h.OptionalLabels) {
+		curriedVec, err = curriedVec.CurryWith(prometheus.Labels{v: GetLabelValue(obj, h.OptionalLabels[v])})
+	}
+	if err != nil {
+		// The possibility of hitting an error here is low, but report it and don't observe the metric in case it happens
+		log.WithError(err).Error("error observing metric")
+	} else {
+		curriedVec.WithLabelValues().Observe(val)
+	}
+}
+
+var _ metricsWithDyanamicLabels = CounterWithDynamicLabels{}
+var _ metricsWithDyanamicLabels = HistogramWithDynamicLabels{}
+
+// GetOptionalClusterTypeLabels reads the AdditionalClusterDeploymentLabels from the metrics config and returns the same
+// as a map
+func GetOptionalClusterTypeLabels(mConfig *metricsconfig.MetricsConfig) map[string]string {
+	var mapClusterTypeLabelToValue = make(map[string]string)
+	// Currently "cluster_type" is a non-optional label for these metrics.
+	// TODO: remove as a part of https://issues.redhat.com/browse/HIVE-2077
+	if _, ok := mConfig.AdditionalClusterDeploymentLabels["cluster_type"]; !ok {
+		mapClusterTypeLabelToValue["cluster_type"] = hivev1.HiveClusterTypeLabel
+	}
+	for k, v := range mConfig.AdditionalClusterDeploymentLabels {
+		mapClusterTypeLabelToValue[k] = v
+	}
+	return mapClusterTypeLabelToValue
+}
+
+// GetSortedKeys gets the keys of a map, sorts them and returns the same.
+func GetSortedKeys(mapToSort map[string]string) []string {
+	var keys []string
```

Comment:
nit: using a `make()` with `len(mapToSort)` will make this func marginally more efficient, as it won't have to reallocate `keys` on every iteration of the loop. (The compiler probably optimizes that away anyway, but...)